### PR TITLE
MH-102 Support file uploads & dynamic upload widget transitions

### PR DESCRIPTION
**Pull request by dsummersl**
This PR:

 * Supports file uploads and viladations as laid out in the ticket.
 * converts all upload steps to ajax calls that are managed by the dashboard.

This PR isn't finished until:

 * [x] A view is shown for HTTP responses that are failures
 * [x] resolve TODOs

### CODE REVIEW

File: myhpom/templates/myhpom/dashboard.html [hunk 1]
```diff
@@ -127,4 +127,63 @@ <h2 class="h6 dashboard-faq__title">
...
+
+    function handleAjax($ajax) {
+        $ajax.done(function (data) {
+            $(ADVANCED_DIRECTIVE_WIDGET_ID).html(data);
+        }).fail(function() {
+            $(ADVANCED_DIRECTIVE_WIDGET_ID).html(AJAX_FAILURE_CONTENT);
+        });
+    }
+
+    function handleFormSubmit() {
```
Comment (nmashton) [inline myhpom/templates/myhpom/dashboard.html]: `event` is unbound here

File: myhpom/templates/myhpom/dashboard.html [hunk 2]
```diff
@@ -127,4 +127,63 @@ <h2 class="h6 dashboard-faq__title">
...
+        event.preventDefault();
+
+        var formData = new FormData();
+        $($(event.target).serializeArray()).each(function (idx, item) {
```
Comment (nmashton) [inline myhpom/templates/myhpom/dashboard.html]: You could probably do a `var $form = $(event.target);` or some such just to cut down on the repetition.

File: myhpom/templates/myhpom/upload/requirements.html [hunk 3]
```diff
@@ -34,20 +34,22 @@ <h3 class="upload-requirements__title">Advance Directive Document Requirements f
     </div>
 </div>
 
+<form id="id_requirements_form" class="upload-requirements__form" method="POST" action="{% url 'myhpom:upload_requirements' %}">
+{% csrf_token %}
+
 <div class="row">
     <div class="col">
         <h4 class="upload-requirements__item-head">My file is valid starting:
             <a class="upload-requirements__help--right" href="" onClick="event.preventDefault();">What is this?</a>
```
Comment (nmashton) [inline myhpom/templates/myhpom/upload/requirements.html:43]: It's admittedly a small thing, but I request that we avoid this by the same token as we avoid inline `style`.

File: myhpom/tests/test_upload_view.py [hunk 4]
```diff
@@ -64,23 +71,85 @@ def test_not_logged_in(self):
...
         user.advancedirective.refresh_from_db()
         self.assertTrue(user.advancedirective.share_with_ehs)
+
+
+class DirectiveUploadRequirementsTestCase(GETMixin, TestCase):
+
+    def setUp(self):
+        self.url = reverse('myhpom:upload_requirements')
+
+    def test__POST_valid_date(self):
```
Comment (nmashton) [inline myhpom/tests/test_upload_view.py]: Why the double underscore?

**Review by nmashton — APPROVED** [review]
I remain baffled by how `event` was ever getting bound, but … it worked then, and it works now! The test logic is clear and everything checks out.

The step on the "sharing form" that's supposed to show the filename currently doesn't, but that will be fixed in my branch.

File: myhpom/templates/myhpom/upload/requirements.html (hunk 3 above)
Comment (dsummersl) [reply myhpom/templates/myhpom/upload/requirements.html:43]: Agreed; I didn't even notice it there 🌴 🤕

File: myhpom/tests/test_upload_view.py (hunk 4 above)
Comment (dsummersl) [reply myhpom/tests/test_upload_view.py]: hrm, dunno - I copied this into place - I'll clean that up.

**Pull request merged**